### PR TITLE
[WIP] Adds `autoProvisioning` feature

### PR DESCRIPTION
**Pull request by ashishranjan738**
This commit adds `autoProvisioning` feature in LSO.

Signed-off-by: Ashish Ranjan <ashishranjan738@gmail.com>

### CODE REVIEW

File: pkg/apis/local/v1alpha1/localvolumeset_types.go [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+	// Partition represents a device-type of partion
+	Partition DeviceType = "Partition"
+)
+
+// DeviceInclusionSpec holds the inclusion filter spec
+type DeviceInclusionSpec struct {
+	// Devices is the list of devices that should be used for automatic detection.
+	// This would be one of the types supported by the local-storage operator. Currently,
+	// the supported types are: disk, part. If the list is empty no devices will be selected.
+	Devices []DeviceType `json:"devices"`
```
Comment (rohantmp) [inline pkg/apis/local/v1alpha1/localvolumeset_types.go:40]: DeviceTypes* looks like this name is a mistake that slipped through in the design too.

File: pkg/controller/localvolumeset/localvolumeset_controller.go [hunk 2]
```diff
@@ -100,54 +113,460 @@ func (r *ReconcileLocalVolumeSet) Reconcile(request reconcile.Request) (reconcil
...
 		}
-
-		// Pod created successfully - don't requeue
-		return reconcile.Result{}, nil
-	} else if err != nil {
-		return reconcile.Result{}, err
+		return err
+	}
+	if !reflect.DeepEqual(ds.Spec, oldDs.Spec) {
+		return r.client.Update(context.TODO(), oldDs)
```
Comment (rohantmp) [inline pkg/controller/localvolumeset/localvolumeset_controller.go:164]: Looks like this was meant to modify `oldDS`, but it's not implemented?

Try https://godoc.org/sigs.k8s.io/controller-runtime/pkg/controller/controllerutil#CreateOrUpdate

File: pkg/controller/localvolumeset/localvolumeset_controller.go [hunk 3]
```diff
@@ -100,54 +113,460 @@ func (r *ReconcileLocalVolumeSet) Reconcile(request reconcile.Request) (reconcil
...
+			},
+		},
+	}
+	addOwner(&ds.ObjectMeta, cr)
+	addOwnerLabels(&ds.ObjectMeta, cr)
+	return ds
+}
+
+// syncProvisionerConfigMap syncs the configmap and returns true if configmap was modified
+func (r *ReconcileLocalVolumeSet) syncProvisionerConfigMap(o *localv1alpha1.LocalVolumeSet) error {
```
Comment (rohantmp) [inline pkg/controller/localvolumeset/localvolumeset_controller.go:234]: this does not handle update

Comment (rohantmp) [reply pkg/controller/localvolumeset/localvolumeset_controller.go:234]: try https://godoc.org/sigs.k8s.io/controller-runtime/pkg/controller/controllerutil#CreateOrUpdate

File: pkg/controller/localvolumeset/localvolumeset_controller.go [hunk 4]
```diff
@@ -100,54 +113,460 @@ func (r *ReconcileLocalVolumeSet) Reconcile(request reconcile.Request) (reconcil
...
+	ds := r.generateLocalProvisionerDaemonset(cr, provisionerConfigMapName)
+	oldDs := &appsv1.DaemonSet{}
+	err := r.client.Get(context.TODO(), types.NamespacedName{Name: ds.Name, Namespace: ds.Namespace}, oldDs)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return r.client.Create(context.TODO(), ds)
+		}
+		return err
+	}
+	if !reflect.DeepEqual(ds.Spec, oldDs.Spec) {
```
Comment (rohantmp) [inline pkg/controller/localvolumeset/localvolumeset_controller.go:318]: same here

Comment (ashishranjan738) [reply pkg/controller/localvolumeset/localvolumeset_controller.go]: updated

File: pkg/controller/localvolumeset/localvolumeset_controller.go (hunk 3 above)
Comment (ashishranjan738) [reply pkg/controller/localvolumeset/localvolumeset_controller.go]: updated

File: pkg/controller/localvolumeset/localvolumeset_controller.go (hunk 2 above)
Comment (ashishranjan738) [reply pkg/controller/localvolumeset/localvolumeset_controller.go]: updated

File: pkg/apis/local/v1alpha1/localvolumeset_types.go (hunk 1 above)
Comment (ashishranjan738) [reply pkg/apis/local/v1alpha1/localvolumeset_types.go:40]: No I think it should be `Devicetype` its a custom data type we created

**Comment by rohantmp** [comment]
please review #110 instead.

**Comment by gnufied** [comment]
Can we close this PR then?

**Comment by ashishranjan738** [comment]
closing this PR since #110 contains the fix

**Pull request closed without merging**